src/proc/constant_evaluator.rs

```diff
@@ -0,0 +1,1710 @@
+                    components: ref src_components,
+                    ty,
+                },
+                &Expression::Literal(_),
+            ) => {
+                let mut components = src_components.clone();
+                for component in &mut components {
+                    *component = self.binary_op(op, *component, right, span)?;
+                }
+                Expression::Compose { ty, components }
+            }
+            (
+                &Expression::Literal(_),
+                &Expression::Compose {
+                    components: ref src_components,
+                    ty,
+                },
+            ) => {
+                let mut components = src_components.clone();
+                for component in &mut components {
+                    *component = self.binary_op(op, left, *component, span)?;
+                }
+                Expression::Compose { ty, components }
+            }
+            _ => return Err(ConstantEvaluatorError::InvalidBinaryOpArgs),
+        };
+
+        self.register_evaluated_expr(expr, span)
+    }
+
+    /// Deep copy `expr` from `expressions` into `self.expressions`.
+    ///
+    /// Return the root of the new copy.
+    ///
+    /// This is used when we're evaluating expressions in a function's
+    /// expression arena that refer to a constant: we need to copy the
+    /// constant's value into the function's arena so we can operate on it.
+    fn copy_from(
+        &mut self,
+        expr: Handle<Expression>,
+        expressions: &Arena<Expression>,
+    ) -> Result<Handle<Expression>, ConstantEvaluatorError> {
+        let span = expressions.get_span(expr);
+        match expressions[expr] {
+            ref expr @ (Expression::Literal(_)
+            | Expression::Constant(_)
+            | Expression::ZeroValue(_)) => self.register_evaluated_expr(expr.clone(), span),
+            Expression::Compose { ty, ref components } => {
+                let mut components = components.clone();
+                for component in &mut components {
+                    *component = self.copy_from(*component, expressions)?;
+                }
+                self.register_evaluated_expr(Expression::Compose { ty, components }, span)
+            }
+            Expression::Splat { size, value } => {
+                let value = self.copy_from(value, expressions)?;
+                self.register_evaluated_expr(Expression::Splat { size, value }, span)
+            }
+            _ => {
+                log::debug!("copy_from: SubexpressionsAreNotConstant");
+                Err(ConstantEvaluatorError::SubexpressionsAreNotConstant)
+            }
+        }
+    }
+
+    fn register_evaluated_expr(
+        &mut self,
+        expr: Expression,
+        span: Span,
+    ) -> Result<Handle<Expression>, ConstantEvaluatorError> {
+        match expr {
+            Expression::Literal(literal) => crate::valid::validate_literal(literal)?,
+            _ => {}
+        }
```

Comment:
In the long run, I'm concerned that we're going to run into a steady stream of problems around error reporting, because it's a fundamental aspect of this design that constant evaluation takes place before other validation, which seems surprising to me. We have to be very careful to only const-evaluate valid expressions, lest we remove errors that need to be reported. And that means that we need to re-implement validation in the constant evaluator.

For now this is fine, but let's keep an eye on this.

Reply:
All that said, this particular call to `validate_literal` is fine because it's checking for a condition that arises only through evaluation. Validating first wouldn't help.

Reply:
The validation code is checking invariants that can't be checked by the type system given the current shape of the IR. Which more specifically for `Expression`s is checking the types of the operands.

The evaluator is operating on strongly typed data (`Literal`s) and is inherently checking the types involved in a given `Expression`.

I'm not sure if I'm missing something but I don't see this being a big issue.